src/popover.ts

```diff
@@ -149,70 +163,132 @@ export function apply() {
     },
   });
 
-  const onClick = (event: Event) => {
-    const target = event.target;
-    if (!(target instanceof Element) || target?.shadowRoot) {
-      return;
-    }
-    const root = target.getRootNode();
-    if (!(root instanceof ShadowRoot || root instanceof Document)) {
+  const popoverTargetAttributesSupportedElementsSelector =
+    ':is(button, input[type="button"], input[type="submit"], input[type="image"], input[type="reset"])';
+
+  const definePopoverTargetElementProperty = (name: string) => {
+    const invokersMap = new WeakMap<Element, Element>();
+    function obj(this: HTMLButtonElement | HTMLInputElement) {
       return;
     }
-    let effectedPopover = closestElement(
-      '[popover]',
-      target,
-    ) as HTMLElement | null;
-    const button = target.closest(
-      '[popovertoggletarget],[popoverhidetarget],[popovershowtarget]',
+    const invokerDescriptor: PropertyDescriptor &
+      ThisType<HTMLButtonElement | HTMLInputElement> = {
+      set(targetElement: unknown) {
+        if (targetElement === null) {
+          this.removeAttribute(name.toLowerCase());
+          invokersMap.delete(this);
+        } else if (!(targetElement instanceof Element)) {
+          throw new TypeError(`${name}Element must be an element or null`);
+        } else {
+          this.setAttribute(name.toLowerCase(), '');
+          invokersMap.set(this, targetElement);
+        }
+      },
+      get() {
+        // If node is not supported, then return null.
+        if (!this.matches(popoverTargetAttributesSupportedElementsSelector)) {
+          return null;
+        }
```

Comment:
We could drop `matches` here, and be a little more explicit:

```suggestion
        // If node is not supported, then return null.
        if (this.localName !== 'button' && this.localName !== 'input') {
          return null;
        }
        
        // If the node is not the correct input type, then return null.
        if (this.localName === 'input' && this.type !== 'reset' && this.type !== 'image' && this.type !== 'button') {
          return null;
        }
```

Matches is fine if you want to keep it, but considering we have the other logic inlined here, it would improve readability IMO

Reply:
agree

Reply:
agree